providers/coordinator.go

```diff
@@ -62,6 +63,33 @@ type RunningProvider struct {
 	lock sync.Mutex
 }
 
+// initialize the heartbeat with the provider
+func (p *RunningProvider) heartbeat() {
+	interval := 2 * time.Second
+	gracePeriod := 3 * time.Second
+	go func() {
+		for !p.isClosed && !p.isShutdown {
+			_, err := p.Plugin.Heartbeat(&pp.HeartbeatReq{
+				Interval: uint64(interval + gracePeriod),
+			})
+			if status, ok := status.FromError(err); ok {
+				if status.Code() == 12 {
+					log.Fatal().
+						Str("plugin-ID", p.ID).
+						Msg("please update the provider plugin for " + p.Name)
+				}
+			}
+			if err != nil {
+				log.Fatal().
```

Comment:
log.Fatal prevents defers from running, meaning we do not get a chance to properly clean up things like the providers